Better memory management

alphaville opened the pull request:
This is an alternative to PR #49; this is still WIP.

## Main Changes

- Introduce `T** m_d_ptrMatrices`, which is a raw pointer to the matrices of the tensor (to replace `pointersToMatrices`)

## TODOs

- [x] New memory management scheme (malloc/free)
- [x] `addAB`: use `gemm` when `nMats=1`
- [x] Slicing: set `m_d_ptrMatrix = nullptr` when slicing with `axis=0` or `axis=1`
- [x] Remove unit test `tensorPointersToMatrices` (method deprecation)
- [x] Get rid of `pointersToMatrices()` in `leastSquaresBatched`
- [x] `CholeskyBatchFactoriser`: get rid of `pointersToMatrices`
- [x] `CholeskyBatchFactoriser`: throw error if `nMats=1`
- [x] Reshaping: new test needed (e.g., reshape + multiply with `addAB`)
- [x] Unit test with reshaping and slicing

## Note

It probably makes sense to always have `m_d_ptrMatrix = nullptr` whenever `nMats=1` and throw an error when the user attempts to use any batched method with `nMats=1`.

## Associated Issues

- Closes #48

alphaville committed:
Create pointer-to-matrices memory (m_d_ptrMatrices)

Allocating and destroying memory
No memory leaks evident atm

alphaville committed:
Initialise m_d_ptrMatrices

Introduce initialisePointersToMatricesData to initialise
m_d_ptrMatrices; appropriate checks for safety.

alphaville committed:
Use initialisePointersToMatricesData to initialise m_d_ptrMatrices

Also in the slice constructor, keep m_d_ptrMatrices=nullptr
when we dont slice along axis=2

alphaville committed:
update addAB to use preallocated memory

use cublasDGemm if nMats=1

alphaville committed:
mirror implementation for DTensor<float>::addAB

alphaville committed:
Get rid of pointersToMatrices() in LS

alphaville committed:
New method DTensor<T>::ptrMatrices() exposes m_d_ptrMatrices

alphaville committed:
Completely got rid of pointersToMatrices()

alphaville committed:
No memory allocation for m_d_ptrMatrices unless nMats > 1

alphaville committed:
ptrMatrices now declared as a const method

alphaville committed:
Update DTensor<T>::reshape

When necessary, memory is reallocated for m_d_ptrMatrices
within reshape

alphaville committed:
Better memory management when reshaping

No need to free/reallocate when we can reuse the already
allocated memory space (when the number of matrices
decreases); no new memory allocation when numMats=1

alphaville committed:
Safe memory allocation in reshape and unit test

alphaville committed:
Test reshape more thoroughly

and format code

ruairimoran commented on include/tensor.cuh:
```diff
@@ -757,12 +825,20 @@ template<typename T>
+
+    if (numMats() > 1) {
+        m_doDestroyPtrMatrices = true;
+        cudaStatus = cudaMalloc(&m_d_ptrMatrices, numMats() * sizeof(T *));
```
why not use `gpuErrChk` for these couple of cuda calls ?

alphaville replied:
Good point. I'll make a few changes there.

ruairimoran commented on include/tensor.cuh:
Line 474 (reshape declaration) needs docs:
/**
 * Reshape tensor.
 * Note: increasing the number of matrices of a tensor involves memory allocation,
 * so is best kept offline.
 */

alphaville replied:
Done

ruairimoran commented on test/testTensor.cu:
```diff
@@ -26,7 +26,7 @@ protected:
 TEMPLATE_WITH_TYPE_T
 void tensorConstructionZero() {
-    DTensor<T> zero(2, 3, 4, true);
+    DTensor <T> zero(2, 3, 4, true);
```
why all the additional spaces in this file ?

alphaville replied:
My stupid code formatter messed up

ruairimoran replied:
hate when that happens

ruairimoran commented on test/testTensor.cu:
```diff
@@ -121,9 +121,9 @@ TEST_F(TensorTest, randomTensorCreation) {
-    DTensor<T> y(DTensor<T>{100, 10, 1000});
+    DTensor <T> zero(2, 3, 4, true);
+    DTensor <T> x(std::move(zero));
+    DTensor <T> y(DTensor < T > {100, 10, 1000});
```
more formatting changes ?

ruairimoran commented on test/testTensor.cu:
```diff
@@ -299,6 +299,50 @@ TEST_F(TensorTest, tensorDeviceCopyTo) {
+
+/* ---------------------------------------
+ * Tensor: Frobenius dot product
+ * --------------------------------------- */
```
Tensor: reshape ?

ruairimoran commented on test/testTensor.cu:
```diff
@@ -421,7 +465,7 @@ void tensorRightGivens(T epsilon) {
 }
 
-TEST_F(TensorTest, tensorRightGivens) {
+TEST_F(TensorTest, tensorRightGivens ) {
```
random whitespace

ruairimoran reviewed (approved):
check comments. Looks great

alphaville committed:
Further unit tests for reshape and docs

Unit test: Slice along axis=2 and reshape
Write documentation for DTensor<T>::reshape
Fix formatting issues in testTensor

ruairimoran reviewed (approved):
nice

alphaville committed:
Better memory management in DTensor<T>::allocateOnDevice

Free all allocated memory if allocation fails
Fix code formatting in testTensor
allocateOnDevice made void

alphaville committed:
DTensor<T>::allocateOnDevice: memento mori
